### PR TITLE
Debugger/SPU: Implement SPU Disassembler

### PR DESCRIPTION
**Pull request by elad335**
Implement a tool that can load SPU code from any PS3 address and disassemble it.

### CODE REVIEW

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 1]
```diff
@@ -1219,7 +1354,7 @@ void debugger_frame::ShowGotoAddressDialog()
 	// -1 from get_pc() turns into 0
 	const u32 pc = cpu ? utils::align<u32>(cpu->get_pc(), 4) : 0;
 	expression_input->setPlaceholderText(QString("0x%1").arg(pc, 16, 16, QChar('0')));
-	expression_input->setFixedWidth(gui::utils::get_label_width(expression_input->placeholderText(), &font));
+	expression_input->setFixedWidth(gui::utils::get_label_width(expression_input->placeholderText(), &font) + 2);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:1357]: Why +2

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 2]
```diff
@@ -327,6 +327,7 @@ void debugger_frame::keyPressEvent(QKeyEvent* event)
 			"\nKey I: Show RSX method detail."
 			"\nKey F10: Perform step-over on instructions. (skip function calls)"
 			"\nKey F11: Perform single-stepping on instructions."
+			"\nKey F12: Show the SPU disassmebler dialog."
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:330]: Typo. In disassembler.
Also, this doesn't match the actual key combo

Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp:330]: I know, its a draft.

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 3]
```diff
@@ -1122,6 +1134,129 @@ void debugger_frame::OnSelectUnit()
...
+	connect(button_cancel, &QAbstractButton::clicked, m_spu_disasm_dialog, &QDialog::reject);
+
+	m_spu_disasm_dialog->move(QCursor::pos());
+	m_spu_disasm_dialog->setAttribute(Qt::WA_DeleteOnClose);
+
+	connect(m_spu_disasm_dialog, &QDialog::finished, this, [this, source_eal, start_pc](int result)
+	{
+		m_spu_disasm_dialog = nullptr;
+
+		if (result == QDialog::Accepted)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:1199]: Why is there a giant scope.
Just early out

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 4]
```diff
@@ -1122,6 +1134,129 @@ void debugger_frame::OnSelectUnit()
...
+		{
+			const u64 start_pc_address = EvaluateExpression(start_pc->text());
+
+			if (start_pc_address > SPU_LS_SIZE - 4)
+			{
+				return;
+			}
+
+			if (const u64 address = EvaluateExpression(source_eal->text()))
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:1208]: Same here

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 5]
```diff
@@ -1122,6 +1134,129 @@ void debugger_frame::OnSelectUnit()
...
+			{
+				cpu_thread* selected = nullptr;
+
+				u32 spu_addr = address;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:1212]: const

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 6]
```diff
@@ -1122,6 +1134,129 @@ void debugger_frame::OnSelectUnit()
...
+				u32 spu_base = start_pc_address;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:1213]: const

File: rpcs3/rpcs3qt/debugger_frame.cpp (hunk 4 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp]: Fixed

**Pull request merged**